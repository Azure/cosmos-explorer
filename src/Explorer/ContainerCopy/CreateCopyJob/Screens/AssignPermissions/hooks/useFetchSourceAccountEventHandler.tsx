import { useCallback, useState } from "react";
import { fetchDatabaseAccount } from "../../../../../../Utils/arm/databaseAccountUtils";
import { useCopyJobContext } from "../../../../Context/CopyJobContext";
import { getAccountDetailsFromResourceId } from "../../../../CopyJobUtils";

const useFetchSourceAccountEventHandler = () => {
  const [loading, setLoading] = useState<boolean>(false);
  const { copyJobState: { source } = {}, setCopyJobState } = useCopyJobContext();
  const onWindowClosed = useCallback(async (errorMessage: string) => {
    try {
      const selectedSourceAccount = source?.account;
      const {
        subscriptionId: sourceSubscriptionId,
        resourceGroup: sourceResourceGroup,
        accountName: sourceAccountName,
      } = getAccountDetailsFromResourceId(selectedSourceAccount?.id);

      setLoading(true);
      const account = await fetchDatabaseAccount(sourceSubscriptionId, sourceResourceGroup, sourceAccountName);
      if (account) {
        setCopyJobState((prevState) => ({
          ...prevState,
          source: { ...prevState.source, account: account },
        }));
      }
    } catch (error) {
      console.error(errorMessage, error);
    } finally {
      setLoading(false);
    }
  }, []);

  return { loading, onWindowClosed };
};

export default useFetchSourceAccountEventHandler;
