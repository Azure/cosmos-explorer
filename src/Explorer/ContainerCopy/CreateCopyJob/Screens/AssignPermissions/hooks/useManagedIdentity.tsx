import { DatabaseAccount } from "Contracts/DataModels";
import { useCallback, useState } from "react";
import { useCopyJobContext } from "../../../../Context/CopyJobContext";
import { getAccountDetailsFromResourceId } from "../../../../CopyJobUtils";

interface UseManagedIdentityUpdaterParams {
    updateIdentityFn: (
        subscriptionId: string,
        resourceGroup?: string,
        accountName?: string
    ) => Promise<DatabaseAccount | undefined>;
}

interface UseManagedIdentityUpdaterReturn {
    loading: boolean;
    handleAddSystemIdentity: () => Promise<void>;
}

const useManagedIdentity = (
    updateIdentityFn: UseManagedIdentityUpdaterParams["updateIdentityFn"]
): UseManagedIdentityUpdaterReturn => {
    const { copyJobState, setCopyJobState } = useCopyJobContext();
    const [loading, setLoading] = useState<boolean>(false);

    const handleAddSystemIdentity = useCallback(async (): Promise<void> => {
        try {
            setLoading(true);
            const selectedTargetAccount = copyJobState?.target?.account;
            const {
                subscriptionId: targetSubscriptionId,
                resourceGroup: targetResourceGroup,
                accountName: targetAccountName
            } = getAccountDetailsFromResourceId(selectedTargetAccount?.id);

            const updatedAccount = await updateIdentityFn(
                targetSubscriptionId,
                targetResourceGroup,
                targetAccountName
            );
            if (updatedAccount) {
                setCopyJobState((prevState) => ({
                    ...prevState,
                    target: { ...prevState.target, account: updatedAccount }
                }));
            }
        } catch (error) {
            console.error("Error enabling system-assigned managed identity:", error);
        } finally {
            setLoading(false);
        }
    }, [copyJobState, updateIdentityFn, setCopyJobState]);

    return { loading, handleAddSystemIdentity };
};

export default useManagedIdentity;