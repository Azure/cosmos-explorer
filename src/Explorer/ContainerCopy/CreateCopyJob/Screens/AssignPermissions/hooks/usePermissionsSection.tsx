import { useEffect, useMemo, useRef, useState } from "react";
import { fetchRoleAssignments, fetchRoleDefinitions, RoleDefinitionType } from "../../../../../../Utils/arm/RbacUtils";
import ContainerCopyMessages from "../../../../ContainerCopyMessages";
import { getAccountDetailsFromResourceId } from "../../../../CopyJobUtils";
import { BackupPolicyType, CopyJobMigrationType, DefaultIdentityType, IdentityType } from "../../../../Enums";
import { CopyJobContextState } from "../../../../Types";
import { useCopyJobPrerequisitesCache } from "../../../Utils/useCopyJobPrerequisitesCache";
import AddManagedIdentity from "../AddManagedIdentity";
import AddReadPermissionToDefaultIdentity from "../AddReadPermissionToDefaultIdentity";
import DefaultManagedIdentity from "../DefaultManagedIdentity";
import OnlineCopyEnabled from "../OnlineCopyEnabled";
import PointInTimeRestore from "../PointInTimeRestore";

export interface PermissionSectionConfig {
  id: string;
  title: string;
  Component: React.ComponentType;
  disabled: boolean;
  completed?: boolean;
  validate?: (state: CopyJobContextState) => boolean | Promise<boolean>;
}

// Section IDs for maintainability
export const SECTION_IDS = {
  addManagedIdentity: "addManagedIdentity",
  defaultManagedIdentity: "defaultManagedIdentity",
  readPermissionAssigned: "readPermissionAssigned",
  pointInTimeRestore: "pointInTimeRestore",
  onlineCopyEnabled: "onlineCopyEnabled",
} as const;

const PERMISSION_SECTIONS_CONFIG: PermissionSectionConfig[] = [
  {
    id: SECTION_IDS.addManagedIdentity,
    title: ContainerCopyMessages.addManagedIdentity.title,
    Component: AddManagedIdentity,
    disabled: true,
    validate: (state: CopyJobContextState) => {
      const targetAccountIdentityType = (state?.target?.account?.identity?.type ?? "").toLowerCase();
      return (
        targetAccountIdentityType === IdentityType.SystemAssigned ||
        targetAccountIdentityType === IdentityType.UserAssigned
      );
    },
  },
  {
    id: SECTION_IDS.defaultManagedIdentity,
    title: ContainerCopyMessages.defaultManagedIdentity.title,
    Component: DefaultManagedIdentity,
    disabled: true,
    validate: (state: CopyJobContextState) => {
      const targetAccountDefaultIdentity = (state?.target?.account?.properties?.defaultIdentity ?? "").toLowerCase();
      return targetAccountDefaultIdentity === DefaultIdentityType.SystemAssignedIdentity;
    },
  },
  {
    id: SECTION_IDS.readPermissionAssigned,
    title: ContainerCopyMessages.readPermissionAssigned.title,
    Component: AddReadPermissionToDefaultIdentity,
    disabled: true,
    validate: async (state: CopyJobContextState) => {
      const principalId = state?.target?.account?.identity?.principalId;
      const selectedSourceAccount = state?.source?.account;
      const {
        subscriptionId: sourceSubscriptionId,
        resourceGroup: sourceResourceGroup,
        accountName: sourceAccountName,
      } = getAccountDetailsFromResourceId(selectedSourceAccount?.id);

      const rolesAssigned = await fetchRoleAssignments(
        sourceSubscriptionId,
        sourceResourceGroup,
        sourceAccountName,
        principalId,
      );

      const roleDefinitions = await fetchRoleDefinitions(rolesAssigned ?? []);
      return checkTargetHasReaderRoleOnSource(roleDefinitions ?? []);
    },
  },
];

const PERMISSION_SECTIONS_FOR_ONLINE_JOBS: PermissionSectionConfig[] = [
  {
    id: SECTION_IDS.pointInTimeRestore,
    title: ContainerCopyMessages.pointInTimeRestore.title,
    Component: PointInTimeRestore,
    disabled: true,
    validate: (state: CopyJobContextState) => {
      const sourceAccountBackupPolicy = state?.source?.account?.properties?.backupPolicy?.type ?? "";
      return sourceAccountBackupPolicy === BackupPolicyType.Continuous;
    },
  },
  {
    id: SECTION_IDS.onlineCopyEnabled,
    title: ContainerCopyMessages.onlineCopyEnabled.title,
    Component: OnlineCopyEnabled,
    disabled: true,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    validate: (_state: CopyJobContextState) => {
      return false;
    },
  },
];

/**
 * Checks if the user has the Reader role based on role definitions.
 */
export function checkTargetHasReaderRoleOnSource(roleDefinitions: RoleDefinitionType[]): boolean {
  return roleDefinitions?.some(
    (role) =>
      role.name === "00000000-0000-0000-0000-000000000001" ||
      role.permissions.some(
        (permission) =>
          permission.dataActions.includes("Microsoft.DocumentDB/databaseAccounts/readMetadata") &&
          permission.dataActions.includes("Microsoft.DocumentDB/databaseAccounts/sqlDatabases/containers/items/read"),
      ),
  );
}

/**
 * Returns the permission sections configuration for the Assign Permissions screen.
 * Memoizes derived values for performance and decouples logic for testability.
 */
const usePermissionSections = (state: CopyJobContextState): PermissionSectionConfig[] => {
  const { validationCache, setValidationCache } = useCopyJobPrerequisitesCache();
  const [permissionSections, setPermissionSections] = useState<PermissionSectionConfig[] | null>(null);
  const isValidatingRef = useRef(false);

  const sectionToValidate = useMemo(() => {
    const baseSections = [...PERMISSION_SECTIONS_CONFIG];
    if (state.migrationType === CopyJobMigrationType.Online) {
      return [...baseSections, ...PERMISSION_SECTIONS_FOR_ONLINE_JOBS];
    }
    return baseSections;
  }, [state.migrationType]);

  const memoizedValidationCache = useMemo(() => {
    if (state.migrationType === CopyJobMigrationType.Offline) {
      validationCache.delete(SECTION_IDS.pointInTimeRestore);
      validationCache.delete(SECTION_IDS.onlineCopyEnabled);
    }
    return validationCache;
  }, [state.migrationType]);

  useEffect(() => {
    const validateSections = async () => {
      if (isValidatingRef.current) {
        return;
      }

      isValidatingRef.current = true;
      const result: PermissionSectionConfig[] = [];
      const newValidationCache = new Map(memoizedValidationCache);

      for (let i = 0; i < sectionToValidate.length; i++) {
        const section = sectionToValidate[i];

        // Check if this section was already validated and passed
        if (newValidationCache.has(section.id) && newValidationCache.get(section.id) === true) {
          result.push({ ...section, completed: true });
          continue;
        }
        // We've reached the first non-cached section - validate it
        if (section.validate) {
          const isValid = await section.validate(state);
          newValidationCache.set(section.id, isValid);
          result.push({ ...section, completed: isValid });
          // Stop validation if current section failed
          if (!isValid) {
            for (let j = i + 1; j < sectionToValidate.length; j++) {
              result.push({ ...sectionToValidate[j], completed: false });
            }
            break;
          }
        } else {
          // Section has no validate method
          newValidationCache.set(section.id, false);
          result.push({ ...section, completed: false });
        }
      }

      setValidationCache(newValidationCache);
      setPermissionSections(result);
      isValidatingRef.current = false;
    };

    validateSections();

    return () => {
      isValidatingRef.current = false;
    };
  }, [state, sectionToValidate]);

  return permissionSections ?? [];
};

export default usePermissionSections;
