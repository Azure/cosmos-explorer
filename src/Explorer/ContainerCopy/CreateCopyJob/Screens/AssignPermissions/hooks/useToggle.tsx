import { useCallback, useState } from "react";

const useToggle = (initialState = false) => {
  const [state, setState] = useState<boolean>(initialState);
  const onToggle = useCallback((_, checked?: boolean) => {
    setState(checked);
  }, []);
  return [state, onToggle] as const;
};

export default useToggle;
