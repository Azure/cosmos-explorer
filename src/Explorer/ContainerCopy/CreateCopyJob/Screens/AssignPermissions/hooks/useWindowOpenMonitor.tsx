import { useEffect, useRef } from "react";

const useWindowOpenMonitor = (url: string, onClose?: () => void, intervalMs = 500) => {
    const intervalRef = useRef<NodeJS.Timeout | null>(null);

    const openWindowAndMonitor = () => {
        const newWindow = window.open(url, '_blank');
        intervalRef.current = setInterval(() => {
            if (newWindow?.closed) {
                clearInterval(intervalRef.current!);
                intervalRef.current = null;
                console.log('New window has been closed!');
                if (onClose) {
                    onClose();
                }
            }
        }, intervalMs);
    };

    useEffect(() => {
        return () => {
            if (intervalRef.current) {
                clearInterval(intervalRef.current);
                intervalRef.current = null;
            }
        };
    }, []);

    return openWindowAndMonitor;
};

export default useWindowOpenMonitor;