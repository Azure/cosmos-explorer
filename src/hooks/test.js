fetch("https://main.documentdb.ext.azure.com/api/cassandra/keys", {
  headers: {
    accept: "*/*",
    "accept-language": "en",
    authorization:
      "Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6Im5PbzNaRHJPRFhFSzFqS1doWHNsSFJfS1hFZyIsImtpZCI6Im5PbzNaRHJPRFhFSzFqS1doWHNsSFJfS1hFZyJ9.eyJhdWQiOiJodHRwczovL21hbmFnZW1lbnQuYXp1cmUuY29tLyIsImlzcyI6Imh0dHBzOi8vc3RzLndpbmRvd3MubmV0LzcyZjk4OGJmLTg2ZjEtNDFhZi05MWFiLTJkN2NkMDExZGI0Ny8iLCJpYXQiOjE2MTU0ODMzNDAsIm5iZiI6MTYxNTQ4MzM0MCwiZXhwIjoxNjE1NDg3MjQwLCJhY3IiOiIxIiwiYWlvIjoiQWFRQVcvOFRBQUFBbjdpdko0WVhhSW1UeFZLbSt2enY5QjZpUUp6U3J6Z1NJZFBvMmcvbDZReHd0Q2JtNmVabTRkTFd5eXFCN3JxN0tRZ1FmaHRESHFGblMwSnpvOVluQStIb2xLMVVJeURDYzByaTRwVDRZbUI3UUF6R1lIYi9IWVVFWHozdkxTeXdiNVpBbDZRQlFWUktKQmh3R1F5MEJFNlk3Q25YZlpBTE5kNFE3Y3oweklvVmdmQTdwV3d6RzlBWXAxM1QyL0wwYjlWUjdCZ2E1QWpFOWw4a1BBVGI5Zz09IiwiYW1yIjpbInJzYSIsIm1mYSJdLCJhcHBpZCI6IjIwM2YxMTQ1LTg1NmEtNDIzMi04M2Q0LWE0MzU2OGZiYTIzZCIsImFwcGlkYWNyIjoiMCIsImRldmljZWlkIjoiMzI3NjNiYjktMDNlNS00ZDBkLTliZmEtZmEyY2U5OGQ1ZGVlIiwiZmFtaWx5X25hbWUiOiJGYXVsa25lciIsImdpdmVuX25hbWUiOiJTdGV2ZSIsImhhc2dyb3VwcyI6InRydWUiLCJpcGFkZHIiOiI0NS4yMi4xMjIuMjIwIiwibmFtZSI6IlN0ZXZlIEZhdWxrbmVyIiwib2lkIjoiN2M4Yjk4ZGItOTA3OC00NGM3LWE5YWItYzJiOGYxOGRiZDM2Iiwib25wcmVtX3NpZCI6IlMtMS01LTIxLTIxMjc1MjExODQtMTYwNDAxMjkyMC0xODg3OTI3NTI3LTMyMjM1ODIxIiwicHVpZCI6IjEwMDM3RkZFQUJDNTk0QjYiLCJyaCI6IjAuQVJvQXY0ajVjdkdHcjBHUnF5MTgwQkhiUjBVUlB5QnFoVEpDZzlTa05XajdvajBhQUprLiIsInNjcCI6InVzZXJfaW1wZXJzb25hdGlvbiIsInN1YiI6InNJV3JwSTFoQVNUWXJoUFVrYWp1NUtQb3Z6SHdzUkdnOTN3U2t1OEs0aW8iLCJ0aWQiOiI3MmY5ODhiZi04NmYxLTQxYWYtOTFhYi0yZDdjZDAxMWRiNDciLCJ1bmlxdWVfbmFtZSI6InN0ZmF1bEBtaWNyb3NvZnQuY29tIiwidXBuIjoic3RmYXVsQG1pY3Jvc29mdC5jb20iLCJ1dGkiOiJKOXdGaXgxN3RFYU5uZzhVckROUEFBIiwidmVyIjoiMS4wIiwieG1zX3RjZHQiOjEyODkyNDE1NDd9.asA8SB1f7W8SjYtbgqpc-wgGMITDcrmhwg1t-ZSBCTm9zVdLPRBTYK8Qbra39Na3RPAIA15CEwLE5OptIN3La1xx3rijKGxvvhlzPZ7bP2-NbpzAiUCDdHK9WxjoB_q-3WrfnFXEoe57fHKXI_o9gFLd1075Xj5fMDNtOIacF2sFOjumOUZKTJWaKnJ9zr7yJjl0VdP1pOLmg72cxZHSuOFGfj3Xp4L5jCfG5r9pz90ECpM3TMI3IHMqAaozC8-SE_RstHqyuBEhaCvKrCTn869XfdijY2EKYk6I1kCXfKOaQ5MKQNxoCzHqQYKNKrcSO3PNDV1aoBLERtwNoCEfLw",
    "cache-control": "no-cache",
    "content-type": "application/x-www-form-urlencoded; charset=UTF-8",
    pragma: "no-cache",
    "sec-ch-ua": '"Google Chrome";v="89", "Chromium";v="89", ";Not A Brand";v="99"',
    "sec-ch-ua-mobile": "?0",
    "sec-fetch-dest": "empty",
    "sec-fetch-mode": "cors",
    "sec-fetch-site": "same-site",
  },
  referrer: "https://cosmos.azure.com/",
  referrerPolicy: "strict-origin-when-cross-origin",
  method: "POST",
  mode: "cors",
  credentials: "include",
});

fetch("https://localhost:1234/proxy/api/cassandra/keys", {
  headers: {
    accept: "*/*",
    "accept-language": "en-US,en;q=0.9",
    authorization:
      "Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6Im5PbzNaRHJPRFhFSzFqS1doWHNsSFJfS1hFZyIsImtpZCI6Im5PbzNaRHJPRFhFSzFqS1doWHNsSFJfS1hFZyJ9.eyJhdWQiOiJodHRwczovL21hbmFnZW1lbnQuY29yZS53aW5kb3dzLm5ldCIsImlzcyI6Imh0dHBzOi8vc3RzLndpbmRvd3MubmV0LzcyZjk4OGJmLTg2ZjEtNDFhZi05MWFiLTJkN2NkMDExZGI0Ny8iLCJpYXQiOjE2MTU0ODI5NDYsIm5iZiI6MTYxNTQ4Mjk0NiwiZXhwIjoxNjE1NTY5NjQ2LCJhaW8iOiJFMlpnWUZpU3NUVG13U1VYaHlmYVdldVhQdGN1QVFBPSIsImFwcGlkIjoiZmQ4NzUzYjAtMDcwNy00ZTMyLTg0ZTktMjUzMmFmODY1ZmI0IiwiYXBwaWRhY3IiOiIxIiwiaWRwIjoiaHR0cHM6Ly9zdHMud2luZG93cy5uZXQvNzJmOTg4YmYtODZmMS00MWFmLTkxYWItMmQ3Y2QwMTFkYjQ3LyIsIm9pZCI6IjY5NGE3NjQ5LTFjYzQtNDk5ZS1iNTIwLTFjMTY4NWNmODViOSIsInJoIjoiMC5BUm9BdjRqNWN2R0dyMEdScXkxODBCSGJSN0JUaF8wSEJ6Sk9oT2tsTXEtR1g3UWFBQUEuIiwic3ViIjoiNjk0YTc2NDktMWNjNC00OTllLWI1MjAtMWMxNjg1Y2Y4NWI5IiwidGlkIjoiNzJmOTg4YmYtODZmMS00MWFmLTkxYWItMmQ3Y2QwMTFkYjQ3IiwidXRpIjoibUhrQVdIcHhZMEswcnJ3amZ1c1ZBQSIsInZlciI6IjEuMCIsInhtc190Y2R0IjoxMjg5MjQxNTQ3fQ.GgOPiDWivSLq5QgXpDkiTWZxVSWDbE7tktorLuPpiUQS_tt5xNc_FLnJur2upbBRGeLw01hxEk4aCi5aA7lwcEx3yjefjUdYYYjxfmjnvfBuAsCZJ4JpOJv-vZ0Ym3HHH036tKv_KNC8bJgNXH_aLxDZIImI8dVgfmzALPFs0NGw36GRG2GKg55LNhzaqalZ4rZFIKas5M3VXqMjGsx3plpI5CnJAAIV5YrxXdAv02yV4Ui0CQL1Nu2XUG4jCRLM8YI0lrBzD6aEbnSY8tA2LFyAa8cA2yKZrnL3NOCe8VL1NfMhIljTbR5lVl5RmfEPxkX067GgAcA2fsC42skulg",
    "cache-control": "no-cache",
    "content-type": "application/x-www-form-urlencoded; charset=UTF-8",
    pragma: "no-cache",
    "sec-ch-ua": '"Google Chrome";v="89", "Chromium";v="89", ";Not A Brand";v="99"',
    "sec-ch-ua-mobile": "?0",
    "sec-fetch-dest": "empty",
    "sec-fetch-mode": "cors",
    "sec-fetch-site": "same-origin",
    "x-requested-with": "XMLHttpRequest",
  },
  referrer: "https://localhost:1234/explorer.html?platform=Portal&disablePortalInitCache",
  referrerPolicy: "strict-origin-when-cross-origin",
  method: "POST",
  mode: "cors",
  credentials: "include",
});
