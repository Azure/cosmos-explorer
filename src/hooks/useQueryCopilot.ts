import { MinimalQueryIterator } from "Common/IteratorUtilities";
import { QueryResults } from "Contracts/ViewModels";
import { guid } from "Explorer/Tables/Utilities";
import create, { UseStore } from "zustand";

interface QueryCopilotState {
  generatedQuery: string;
  likeQuery: boolean;
  userPrompt: string;
  showFeedbackModal: boolean;
  hideFeedbackModalForLikedQueries: boolean;
  correlationId: string;
  query: string;
  selectedQuery: string;
  isGeneratingQuery: boolean;
  isExecuting: boolean;
  dislikeQuery: boolean | undefined;
  showCallout: boolean;
  showSamplePrompts: boolean;
  queryIterator: MinimalQueryIterator | undefined;
  queryResults: QueryResults | undefined;
  errorMessage: string;
  isSamplePromptsOpen: boolean;
  showDeletePopup: boolean;
  showFeedbackBar: boolean;
  showCopyPopup: boolean;
  showErrorMessageBar: boolean;
  generatedQueryComments: string;

  openFeedbackModal: (generatedQuery: string, likeQuery: boolean, userPrompt: string) => void;
  closeFeedbackModal: () => void;
  setHideFeedbackModalForLikedQueries: (hideFeedbackModalForLikedQueries: boolean) => void;
  refreshCorrelationId: () => void;
  setUserPrompt: (userPrompt: string) => void;
  setQuery: (query: string) => void;
  setGeneratedQuery: (generatedQuery: string) => void;
  setSelectedQuery: (selectedQuery: string) => void;
  setIsGeneratingQuery: (isGeneratingQuery: boolean) => void;
  setIsExecuting: (isExecuting: boolean) => void;
  setLikeQuery: (likeQuery: boolean) => void;
  setDislikeQuery: (dislikeQuery: boolean | undefined) => void;
  setShowCallout: (showCallout: boolean) => void;
  setShowSamplePrompts: (showSamplePrompts: boolean) => void;
  setQueryIterator: (queryIterator: MinimalQueryIterator | undefined) => void;
  setQueryResults: (queryResults: QueryResults | undefined) => void;
  setErrorMessage: (errorMessage: string) => void;
  setIsSamplePromptsOpen: (isSamplePromptsOpen: boolean) => void;
  setShowDeletePopup: (showDeletePopup: boolean) => void;
  setShowFeedbackBar: (showFeedbackBar: boolean) => void;
  setshowCopyPopup: (showCopyPopup: boolean) => void;
  setShowErrorMessageBar: (showErrorMessageBar: boolean) => void;
  setGeneratedQueryComments: (generatedQueryComments: string) => void;
  resetQueryCopilotStates: () => void;
}

type QueryCopilotStore = UseStore<QueryCopilotState>;

export const useQueryCopilot: QueryCopilotStore = create((set) => ({
  generatedQuery: "",
  likeQuery: false,
  userPrompt: "",
  showFeedbackModal: false,
  hideFeedbackModalForLikedQueries: false,
  correlationId: "",
  query: "",
  selectedQuery: "",
  isGeneratingQuery: false,
  isExecuting: false,
  dislikeQuery: undefined,
  showCallout: false,
  showSamplePrompts: false,
  queryIterator: undefined,
  queryResults: undefined,
  errorMessage: "",
  isSamplePromptsOpen: false,
  showDeletePopup: false,
  showFeedbackBar: false,
  showCopyPopup: false,
  showErrorMessageBar: false,
  generatedQueryComments: "",

  openFeedbackModal: (generatedQuery: string, likeQuery: boolean, userPrompt: string) =>
    set({ generatedQuery, likeQuery, userPrompt, showFeedbackModal: true }),
  closeFeedbackModal: () => set({ generatedQuery: "", likeQuery: false, userPrompt: "", showFeedbackModal: false }),
  setHideFeedbackModalForLikedQueries: (hideFeedbackModalForLikedQueries: boolean) =>
    set({ hideFeedbackModalForLikedQueries }),
  refreshCorrelationId: () => set({ correlationId: guid() }),
  setUserPrompt: (userPrompt: string) => set({ userPrompt }),
  setQuery: (query: string) => set({ query }),
  setGeneratedQuery: (generatedQuery: string) => set({ generatedQuery }),
  setSelectedQuery: (selectedQuery: string) => set({ selectedQuery }),
  setIsGeneratingQuery: (isGeneratingQuery: boolean) => set({ isGeneratingQuery }),
  setIsExecuting: (isExecuting: boolean) => set({ isExecuting }),
  setLikeQuery: (likeQuery: boolean) => set({ likeQuery }),
  setDislikeQuery: (dislikeQuery: boolean | undefined) => set({ dislikeQuery }),
  setShowCallout: (showCallout: boolean) => set({ showCallout }),
  setShowSamplePrompts: (showSamplePrompts: boolean) => set({ showSamplePrompts }),
  setQueryIterator: (queryIterator: MinimalQueryIterator | undefined) => set({ queryIterator }),
  setQueryResults: (queryResults: QueryResults | undefined) => set({ queryResults }),
  setErrorMessage: (errorMessage: string) => set({ errorMessage }),
  setIsSamplePromptsOpen: (isSamplePromptsOpen: boolean) => set({ isSamplePromptsOpen }),
  setShowDeletePopup: (showDeletePopup: boolean) => set({ showDeletePopup }),
  setShowFeedbackBar: (showFeedbackBar: boolean) => set({ showFeedbackBar }),
  setshowCopyPopup: (showCopyPopup: boolean) => set({ showCopyPopup }),
  setShowErrorMessageBar: (showErrorMessageBar: boolean) => set({ showErrorMessageBar }),
  setGeneratedQueryComments: (generatedQueryComments: string) => set({ generatedQueryComments }),

  resetQueryCopilotStates: () => {
    set((state) => ({
      ...state,
      generatedQuery: "",
      likeQuery: false,
      userPrompt: "",
      showFeedbackModal: false,
      hideFeedbackModalForLikedQueries: false,
      correlationId: "",
      query: "",
      selectedQuery: "",
      isGeneratingQuery: false,
      isExecuting: false,
      dislikeQuery: undefined,
      showCallout: false,
      showSamplePrompts: false,
      queryIterator: undefined,
      queryResults: undefined,
      errorMessage: "",
      isSamplePromptsOpen: false,
      showDeletePopup: false,
      showFeedbackBar: false,
      showCopyPopup: false,
      showErrorMessageBar: false,
      generatedQueryComments: "",
    }));
  },
}));
