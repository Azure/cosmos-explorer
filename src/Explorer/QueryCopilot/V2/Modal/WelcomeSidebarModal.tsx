import { Image, Link, PrimaryButton, Stack, Text } from "@fluentui/react";
import { useQueryCopilot } from "hooks/useQueryCopilot";
import React from "react";
import Database from "../../../../../images/CopilotDatabase.svg";
import Flash from "../../../../../images/CopilotFlash.svg";
import CopilotSidebarWelcomeIllustration from "../../../../../images/CopilotSidebarWelcomeIllustration.svg";
import Thumb from "../../../../../images/CopilotThumb.svg";

export const WelcomeSidebarModal: React.FC = (): JSX.Element => {
  const { showWelcomeSidebar, setShowWelcomeSidebar } = useQueryCopilot();

  const hideModal = (): void => {
    setShowWelcomeSidebar(false);
    window.localStorage.setItem("showWelcomeSidebar", "false");
  };

  React.useEffect(() => {
    const showWelcomeSidebar = window.localStorage.getItem("showWelcomeSidebar");
    setShowWelcomeSidebar(showWelcomeSidebar && showWelcomeSidebar === "false" ? false : true);
  }, []);

  return (
    showWelcomeSidebar && (
      <Stack
        style={{
          width: "100%",
          height: "100%",
          overflow: "auto",
          backgroundColor: "#FAFAFA",
          flex: "1 0 100%",
        }}
      >
        <div
          style={{
            margin: "20px 10px",
            padding: "20px",
            maxHeight: "100%",
            boxSizing: "border-box",
            borderRadius: "20px",
            backgroundColor: "white",
          }}
        >
          <Stack horizontalAlign="center" verticalAlign="center">
            <Image src={CopilotSidebarWelcomeIllustration} />
          </Stack>

          <Stack>
            <Stack.Item align="center" style={{ marginBottom: "10px" }}>
              <Text className="title bold">Welcome to Copilot in CosmosDB</Text>
            </Stack.Item>
            <Stack.Item style={{ marginBottom: "15px" }}>
              <Stack>
                <Stack horizontal>
                  <Stack.Item align="start">
                    <Image src={Flash} />
                  </Stack.Item>
                  <Stack.Item align="center" style={{ marginLeft: "10px" }}>
                    <Text style={{ fontWeight: 600 }}>
                      Let copilot do the work for you
                      <br />
                    </Text>
                  </Stack.Item>
                </Stack>
                <Stack.Item style={{ textAlign: "start", marginLeft: "25px" }}>
                  <Text>
                    Ask Copilot to generate a query by describing the query in your words.
                    <br />
                    <Link target="_blank" href="https://aka.ms/cdb-copilot-learn-more">
                      Learn more
                    </Link>
                  </Text>
                </Stack.Item>
              </Stack>
            </Stack.Item>
            <Stack.Item style={{ marginBottom: "15px" }}>
              <Stack>
                <Stack horizontal>
                  <Stack.Item align="start">
                    <Image src={Thumb} />
                  </Stack.Item>
                  <Stack.Item align="center" style={{ marginLeft: "10px" }}>
                    <Text style={{ fontWeight: 600 }}>
                      Use your judgement
                      <br />
                    </Text>
                  </Stack.Item>
                </Stack>
                <Stack.Item style={{ textAlign: "start", marginLeft: "25px" }}>
                  <Text>
                    AI-generated content can have mistakes. Make sure itâ€™s accurate and appropriate before using it.
                    <br />
                    <Link target="_blank" href="https://aka.ms/cdb-copilot-preview-terms">
                      Read preview terms
                    </Link>
                  </Text>
                </Stack.Item>
              </Stack>
            </Stack.Item>
            <Stack.Item style={{ marginBottom: "15px" }}>
              <Stack>
                <Stack horizontal>
                  <Stack.Item align="start">
                    <Image src={Database} />
                  </Stack.Item>
                  <Stack.Item align="center" style={{ marginLeft: "10px" }}>
                    <Text style={{ fontWeight: 600 }}>
                      Copilot currently works only a sample database
                      <br />
                    </Text>
                  </Stack.Item>
                </Stack>
                <Stack.Item style={{ textAlign: "start", marginLeft: "25px" }}>
                  <Text>
                    Copilot is setup on a sample database we have configured for you at no cost
                    <br />
                    <Link target="_blank" href="https://aka.ms/cdb-copilot-learn-more">
                      Learn more
                    </Link>
                  </Text>
                </Stack.Item>
              </Stack>
            </Stack.Item>
          </Stack>

          <Stack>
            <Stack.Item align="center">
              <PrimaryButton style={{ width: "224px", height: "32px" }} onClick={hideModal}>
                Get Started
              </PrimaryButton>
            </Stack.Item>
          </Stack>
        </div>
      </Stack>
    )
  );
};
