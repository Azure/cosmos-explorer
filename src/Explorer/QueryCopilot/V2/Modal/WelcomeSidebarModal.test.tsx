import { PrimaryButton } from "@fluentui/react";
import { shallow } from "enzyme";
import { useQueryCopilot } from "hooks/useQueryCopilot";
import { withHooks } from "jest-react-hooks-shallow";
import React from "react";
import { WelcomeSidebarModal } from "./WelcomeSidebarModal";

describe("WelcomeSidebarModal snapshot test", () => {
  it("should close on button click ", () => {
    withHooks(() => {
      const wrapper = shallow(<WelcomeSidebarModal />);
      const spy = jest.spyOn(localStorage, "setItem");
      spy.mockClear();

      const button = wrapper.find(PrimaryButton).first();
      button.simulate("click", {});

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenLastCalledWith("showWelcomeSidebar", "false");
      expect(useQueryCopilot.getState().showWelcomeSidebar).toBeFalsy();
      expect(wrapper).toMatchSnapshot();
    });
  });
  it("should not reneder with local  storage key", () => {
    withHooks(() => {
      window.localStorage.setItem("showWelcomeSidebar", "false");
      const wrapper = shallow(<WelcomeSidebarModal />);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
