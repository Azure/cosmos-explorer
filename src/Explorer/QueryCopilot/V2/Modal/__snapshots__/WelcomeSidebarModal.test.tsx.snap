// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Welcome Sidebar Modal snapshot test should close on button click 1`] = `""`;

exports[`Welcome Sidebar Modal snapshot test should not reneder with local  storage key 1`] = `""`;
